### ISSUE TITLE
Optimize DeltaManager.getDeltas(): consider changing batch size

### ISSUE THREAD

**Issue by vladsud** [microsoft/FluidFramework#5211]
Currently it requests 2K ops a piece, and waits for them to come back before asking for next batch.
This is not very efficient if we have a lot of ops to catch up.
We should consider one of the approaches (or both):
1. Do multiple request in parallel for 2K chunks
2. Not limit it to 2K size and let storage dictate chunk size.

Note that # 2 would make it harder / impossible to do # 1 in DeltaManager, but if # 2 is only about pushing this logic into driver (but keeping wire protocol the same), then under covers driver may consider either of these 2 approaches.

**Comment by vladsud** [comment]
Here is the query that gives us some data:

Office_Fluid_FluidRuntime_Performance
| where Data_eventName contains "fluid:telemetry:DeltaManager:GetDeltas_"
| where Data_eventName contains "_end"
| where Data_deltasRetrievedTotal > 2000
| summarize count() by Data_requests

It captures 3500 total cases that resulted in ~38K requests.
For 1.1M of loads. So we can improve perf of about 3.4% of sessions if we invest here.
That said, the tail is long (there are over 1200 cases where we fetched over 10K ops), so it's worth improving it.
Worth pointing out that most of these issues are for GetDeltas_DocumentOpen_end, i.e. likely using cached snapshots in very busy document, so there is also clear perceived problem for the user.

**Comment by vladsud** [comment]
Note: 100K of messages (87Mb in size) compresses to 4Mb.
I believe the first step here is
- ensure we are getting gzipped payload.
- Increase payload size way over 2K of messages.
- it would be great to see if we can get and report compressed payload size in all of our telemetry (not just ops).

**Comment by vladsud** [comment]
One important data point:
Out of 1220 requests that returned over 10K ops, only in 200 cases we knew about the end ('to' argument).
While we are lacking info about end of file when we start fetching, we likely have that info after first packet arrives due to connection to webSocket. So with this info, we could switch to many requests in parallel

**Comment by vladsud** [comment]
More stats:

Chapter 1 doc:
100K ops is between 76-92Mb
Compressed, take between 3.8-5.6Mb

Chapter2:
81-88Mb
Compressed: 2.8-7.1Mb

So with these numbers, we could go up to 10K ops per request easily.

**Comment by vladsud** [comment]
Here is how long (days) it takes to accumulate 100K ops:

Chapter 2:
20.5
48.6
42.0
24.7
 7.1
 5.9

Chapter 1:
20.4
55.6
42.2
27.7

So with exception of last 2 chunks in Chapter 2 case, we have 20+ days per 100K ops, or less than 5K ops per day.
If we assuming all the issues in https://github.com/microsoft/FluidFramework/issues/5288 are addressed, this should bring it to less than 2.5K ops / day.

As for the peak, we see steady 500ms per op during 25K ops (almost 4 hours).
14.4K of that is lastEditDetailsKey, 5.8K is noopkey - all the known bugs tracked in issue above, we would get to 30K ops / day, if same rate continued through the day. So I think we should expect **~10K ops / day** for similar busy documents for peaks, and down to **2-3K ops / day** on average day in active lifetime of similar documents.

**Comment by vladsud** [comment]
Chapter 2 document - downloading:
1. in 2K chunks:
   - **823** seconds to retrieve 713452 ops in 358 requests
   - **818** seconds to retrieve 713452 ops in 358 requests
2. in 5K chunks
3. in 10K chunks

**Comment by vladsud** [comment]
One more data point. Excluding Chapter 2 doc (which has summaries broken right now), here are percentiles for fluid:telemetry:DeltaManager:GetDeltas_DocumentOpen_end event, Data_deltasRetrievedTotal (looking only at Data_duration > 1000 to attempt to remove cases that were satisfied from ops that came with summary):

50th percentile:    15.5
90th percentile:     767
95th percentile:   2399
98th percentile:  10575
99th percentile:  23486

**Comment by vladsud** [comment]
And in prod, duration of single request of 2K ops is
50th percentile: 2.6 s
90th percentile: 18s
95th percentile: 37s
98th percentile: 959s

This data, though, likely includes (at the right end) substantial amount of time waiting to get back online, while calling token handler. So they do not represent time for actual fetch - this is coming with https://github.com/microsoft/FluidFramework/pull/5273

**Comment by vladsud** [comment]
Note: we learn today about latest seq number only from web socket connection.
It would be great to learn it sooner - from snapshot itself (including when we loaded from cached snapshot, but got newest snapshot to populate cache). We also record latest known op to cache, so we could retrieve that too.
https://onedrive.visualstudio.com/SPIN/_workitems/edit/1068267/

**Comment by vladsud** [comment]
Forked parallelization work into separate issue (above)

**Comment by vladsud** [comment]
Moved to 5K batch size, closing this issue
